### PR TITLE
HV-1395 Support OptionalInt/OptionalDouble/OptionalLong in the annotation processor

### PR DESCRIPTION
**Pull request by marko-bekhta**
- https://hibernate.atlassian.net/browse/HV-1395

### CODE REVIEW

**Comment by gsmet** [comment]
@marko-bekhta I force pushed some very minor changes and will add a few comments inline.

File: annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java [hunk 1]
```diff
@@ -60,25 +69,60 @@ public TypeCheck(ConstraintHelper constraintHelper) {
...
 	}
 
+	private boolean isAnnotationAllowedForType(AnnotationMirror annotation, TypeMirror type) {
+		return ConstraintCheckResult.ALLOWED.equals( constraintHelper.checkConstraint( annotation.getAnnotationType(), type ) );
+	}
+
+	private boolean usesUnwrapping(AnnotationMirror annotationMirror, TypeMirror typeMirror) {
+		//need to check if this annotation is not on one of the types that are automatically unwrapped:
+		if ( constraintHelper.isSupportedForUnwrappingByDefault( getQualifiedName( typeMirror ) ) ) {
+			return true;
```
Comment (gsmet) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java]: We have to take into account the case where it's unwrapped by default BUT we have the `Skip` payload. In this case, it doesn't use unwrapping.

File: annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java [hunk 2]
```diff
@@ -60,25 +69,60 @@ public TypeCheck(ConstraintHelper constraintHelper) {
...
+	}
+
+	private TypeMirror getUnwrappedType(TypeMirror type) {
+		Optional<TypeMirror> optional = constraintHelper.getUnwrappedToByDefault( getQualifiedName( type ) );
+		if ( optional.isPresent() ) {
+			return optional.get();
+		}
+		else {
+			//TODO: need to find a way to check for unwrapping
+			return type;
```
Comment (gsmet) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java]: Thinking out loud, could we allow to declare the types in the annotation processor configuration (thinking about Maven configuration here)?

Something like:
```xml
<containerTypesUnwrappedByDefault>
    <containerTypeUnwrappedByDefault>
        <containerType>...</containerType>
        <unwrappedType>...</unwrappedType>
    </containerTypeUnwrappedByDefault>
</containerTypesUnwrappedByDefault>
```

I mean, one way or another, we need a way to declare them.

Note: it's not blocking for merging this PR, just thinking out loud. We can do it in a second step (if even feasible).

Comment (gsmet) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java]: By the way, in the case we don't have the unwrapped type, I think we should throw a warning and avoid validating anything after that. Don't know if it's complicated to do that in the current execution flow but I think it's the best we can do.

Note that even if we allow to declare the types, I think this warning thing would still be useful.

Comment (marko-bekhta) [inline annotation-processor/src/main/java/org/hibernate/validator/ap/internal/checks/TypeCheck.java]: Yes, I think something like that would be nice to have. That way it'll be possible to register custom extractors. And we can add the ones that we already have somewhere in the code, so they will not be needed in the properties.

**Comment by gsmet** [comment]
Rebased and merged, thanks!

@marko-bekhta could you create an issue about this configuration idea so that we don't forget about it? Thanks!

**Comment by marko-bekhta** [comment]
@gsmet - https://hibernate.atlassian.net/browse/HV-1465 - done :)

**Pull request closed without merging**